# Homework 04:  More Tools for the Toolbox: Null Values, Categorical Features, and Cross Validation

In this homework, we are going to add three tools to your toolbox which will be essential when you work with real datasets:
1. What do we do with null-values?
2. How do we deal with non-numeric features?
3. What validation strategy provides the best estimate of the final testing score?

For (1), we'll explore several ways of dealing with null values:
- Removing columns with too many null values,
- Imputing values for missing categorical labels using the "most frequent" category strategy, and
- Imputing values for missing numeric values using the median. 


For (2), we'll use ordinal encoding to replace categorical labels with floats.

For (3), we'll try three different cross-validation strategies:

- 5-Fold CV,
- Repeated 5-Fold CV, and
- Leave one out CV, 

and see which comes closest to estimating the final testing MSE. 


#### Grading: There are nine (9) answers to provide, each worth 5 points.  (You get 5 points for free.)
 

In [1]:
# Useful imports

import os
import kagglehub
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, cross_val_score, RepeatedKFold, LeaveOneOut
from sklearn.linear_model    import LinearRegression
from sklearn.preprocessing   import OrdinalEncoder, OneHotEncoder  
from sklearn.impute          import SimpleImputer
from sklearn.metrics         import mean_squared_error, r2_score
from tqdm                    import tqdm



### Load the Ames Housing Dataset

For a description of the features of this dataset, see the **Appendix**. 

In [2]:
# Download the latest version of the dataset
path = kagglehub.dataset_download("shashanknecrothapa/ames-housing-dataset")

print("Path to dataset files:", path)

# Construct the full path to the CSV file (update the file name if necessary)
csv_file = os.path.join(path, "AmesHousing.csv")

# Read the dataset into a DataFrame
df = pd.read_csv(csv_file)

Path to dataset files: /Users/waynesnyder/.cache/kagglehub/datasets/shashanknecrothapa/ames-housing-dataset/versions/1


In [3]:
# Print the first few rows of the DataFrame
df.head()

Order        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0      1  526301100           20        RL         141.0     31770   Pave   
1      2  526350040           20        RH          80.0     11622   Pave   
2      3  526351010           20        RL          81.0     14267   Pave   
3      4  526353030           20        RL          93.0     11160   Pave   
4      5  527105010           60        RL          74.0     13830   Pave   

  Alley Lot Shape Land Contour  ... Pool Area Pool QC  Fence Misc Feature  \
0   NaN       IR1          Lvl  ...         0     NaN    NaN          NaN   
1   NaN       Reg          Lvl  ...         0     NaN  MnPrv          NaN   
2   NaN       IR1          Lvl  ...         0     NaN    NaN         Gar2   
3   NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   
4   NaN       IR1          Lvl  ...         0     NaN  MnPrv          NaN   

  Misc Val Mo Sold Yr Sold Sale Type  Sale Condition  SalePrice  
0        0       5    2010       WD           Normal     215000  
1        0       6    2010       WD           Normal     105000  
2    12500       6    2010       WD           Normal     172000  
3        0       4    2010       WD           Normal     244000  
4        0       3    2010       WD           Normal     189900  

[5 rows x 82 columns]

In [4]:
# Uncomment this to see the listing of features

# df.info()

In [5]:
# Uncomment this to see the feature histograms

# print("Feature Histograms")
# df.hist(figsize=(15, 13), bins=30)  # Adjust figure size and number of bins
# plt.tight_layout()  # Adjust spacing to prevent overlap
# plt.show()

### Data Preprocessing

First, let's remove the features that are clearly not useful for regression.

In [6]:
df_clean = df.drop(columns=['Order','PID'])
df_clean.head()

MS SubClass MS Zoning  Lot Frontage  Lot Area Street Alley Lot Shape  \
0           20        RL         141.0     31770   Pave   NaN       IR1   
1           20        RH          80.0     11622   Pave   NaN       Reg   
2           20        RL          81.0     14267   Pave   NaN       IR1   
3           20        RL          93.0     11160   Pave   NaN       Reg   
4           60        RL          74.0     13830   Pave   NaN       IR1   

  Land Contour Utilities Lot Config  ... Pool Area Pool QC  Fence  \
0          Lvl    AllPub     Corner  ...         0     NaN    NaN   
1          Lvl    AllPub     Inside  ...         0     NaN  MnPrv   
2          Lvl    AllPub     Corner  ...         0     NaN    NaN   
3          Lvl    AllPub     Corner  ...         0     NaN    NaN   
4          Lvl    AllPub     Inside  ...         0     NaN  MnPrv   

  Misc Feature Misc Val Mo Sold  Yr Sold  Sale Type  Sale Condition  SalePrice  
0          NaN        0       5     2010        WD           Normal     215000  
1          NaN        0       6     2010        WD           Normal     105000  
2         Gar2    12500       6     2010        WD           Normal     172000  
3          NaN        0       4     2010        WD           Normal     244000  
4          NaN        0       3     2010        WD           Normal     189900  

[5 rows x 80 columns]

### Problem One: Dealing with Null Values

There are basically two strategies for dealing with missing (null or `NaN`) values:
- Get them out of your dataset by **removing** features and/or samples containing too many nulls.
- **Impute** values by replacing nulls with the mean, median, or other "neutral" value computed from the feature.

**Note:** It is also possible to impute values using more advanced techniques such as mode imputation, forward/backward fill, or predictive modeling (e.g., KNN or regression-based imputation). These techniques might be useful when you start to work on your projecct. 

**First we will explore how many null values occur in each feature.**

In [7]:
# This function will list how many nulls occur in which features

def show_null_counts_features(df):
    # Count the nulls and calculate the %
    count_nulls = df.isnull().sum()
    df_nulls = (df.isnull().mean() * 100).round(2)
    
    # Determine if the column is numeric or non-numeric
    feature_types = df.dtypes.apply(lambda x: 'Numeric' if np.issubdtype(x, np.number) else 'Categorical')
    
    # Filter out the columns with missing values and sort them in descending order
    missing_data = pd.DataFrame({
        'Feature': count_nulls[count_nulls > 0].index,
        '# Null Values': count_nulls[count_nulls > 0].values, 
        'Null %': df_nulls[df_nulls > 0].values,
        'Type': feature_types[count_nulls > 0].values
    }).sort_values(by='Null %', ascending=False)
    
    print(f'The dataset contains {len(df)} samples.\n')

    if (len(missing_data) == 0):
        print("There are no null values in the dataset!")
    else:
        # Print null value stats
        print('Feature Name    # Nulls      Null %    Type')
        print('------------    -------      ------    ----')
        for index, row in missing_data.iterrows():
            print(f"{row['Feature']:<15} {row['# Null Values']:<12} {row['Null %']:.2f}%   {row['Type']}")

show_null_counts_features(df_clean)

The dataset contains 2930 samples.

Feature Name    # Nulls      Null %    Type
------------    -------      ------    ----
Pool QC         2917         99.56%   Categorical
Misc Feature    2824         96.38%   Categorical
Alley           2732         93.24%   Categorical
Fence           2358         80.48%   Categorical
Mas Vnr Type    1775         60.58%   Categorical
Fireplace Qu    1422         48.53%   Categorical
Lot Frontage    490          16.72%   Numeric
Garage Cond     159          5.43%   Categorical
Garage Qual     159          5.43%   Categorical
Garage Finish   159          5.43%   Categorical
Garage Yr Blt   159          5.43%   Numeric
Garage Type     157          5.36%   Categorical
Bsmt Exposure   83           2.83%   Categorical
BsmtFin Type 2  81           2.76%   Categorical
Bsmt Cond       80           2.73%   Categorical
Bsmt Qual       80           2.73%   Categorical
BsmtFin Type 1  80           2.73%   Categorical
Mas Vnr Area    23           0.78%   Numeric

### Part A

Clearly, some of these features are not very informative! Let's drop the worst offenders!

**Fill in your code after the comments below to drop any features with more than `max_nulls` null values.**


In [8]:
max_nulls = 500      # we will drop any features with more than max_nulls missing values

# Count null values per column

count_nulls = df_clean.isnull().sum()     # This returns a Series

# to check

print(count_nulls.to_string())       # to_string makes sure to see the whole thing

# Filter out columns where null count exceeds max_nulls

cols_to_drop = count_nulls[count_nulls > max_nulls].index


# Drop the columns

df_clean = df_clean.drop(columns=cols_to_drop)

# Verify they were removed

show_null_counts_features(df_clean)

MS SubClass           0
MS Zoning             0
Lot Frontage        490
Lot Area              0
Street                0
Alley              2732
Lot Shape             0
Land Contour          0
Utilities             0
Lot Config            0
Land Slope            0
Neighborhood          0
Condition 1           0
Condition 2           0
Bldg Type             0
House Style           0
Overall Qual          0
Overall Cond          0
Year Built            0
Year Remod/Add        0
Roof Style            0
Roof Matl             0
Exterior 1st          0
Exterior 2nd          0
Mas Vnr Type       1775
Mas Vnr Area         23
Exter Qual            0
Exter Cond            0
Foundation            0
Bsmt Qual            80
Bsmt Cond            80
Bsmt Exposure        83
BsmtFin Type 1       80
BsmtFin SF 1          1
BsmtFin Type 2       81
BsmtFin SF 2          1
Bsmt Unf SF           1
Total Bsmt SF         1
Heating               0
Heating QC            0
Central Air           0
Electrical      

In [9]:
# Set this variable to the number of cells that were dropped

a1a = len(cols_to_drop)

print(f"a1a = {a1a}")                                   # Don't delete or change this line, it is needed by the auto-grader

a1a = 6


### Part B:  Feature Transformations for Imputing Null Values

Now let's perform the following feature transformations:

- For categorical features, we'll replace null values with the most frequent category in that column
- For numeric features, we'll replace nulls with the median for that feature 


This is very simple to do with a couple of lines of Python, but naturally we want to use `sklearn` functions whenever we can, so we'll use ` SimpleImputer`.

**Go read the doc page for `SimpleImputer` before proceeding.**

### Imputing Categorical Features using the Most Frequent Strategy

In the next cell you see how easy it is to examine the categories. **Take a moment and explore several of the categorical features.**  In this dataset, most of them are skewed, with a clear "most favorite" category. 
(If the feature values are not skewed, then you could change these to a new category "Unknown".)

In [10]:
df_clean['Garage Qual'].value_counts(dropna=False)

Garage Qual
TA     2615
NaN     159
Fa      124
Gd       24
Po        5
Ex        3
Name: count, dtype: int64

Before proceeding, let's get lists of the two types of features.

In [11]:
# Identify categorical and numeric features

categorical_features = df_clean.select_dtypes(exclude=['number']).columns.tolist()
numeric_features     = df_clean.select_dtypes(include=['number']).columns.tolist()

# Print results if you want
# print("Categorical Features:", categorical_features)
# print("Numeric Features:", numeric_features)

**Now you must impute all the categorical features using `SimpleImputer` with the `most_frequent` strategy.**

In [12]:
# First make a copy of the cleaned dataset

df_imputed = df_clean.copy()

# Impute categorical columns (using most frequent category)

cat_imputer = SimpleImputer(strategy="most_frequent")

df_imputed[categorical_features] = cat_imputer.fit_transform(df_imputed[categorical_features])

# Verify: only numeric features should appear

show_null_counts_features(df_imputed)

The dataset contains 2930 samples.

Feature Name    # Nulls      Null %    Type
------------    -------      ------    ----
Lot Frontage    490          16.72%   Numeric
Garage Yr Blt   159          5.43%   Numeric
Mas Vnr Area    23           0.78%   Numeric
Bsmt Full Bath  2            0.07%   Numeric
Bsmt Half Bath  2            0.07%   Numeric
BsmtFin SF 1    1            0.03%   Numeric
BsmtFin SF 2    1            0.03%   Numeric
Bsmt Unf SF     1            0.03%   Numeric
Total Bsmt SF   1            0.03%   Numeric
Garage Cars     1            0.03%   Numeric
Garage Area     1            0.03%   Numeric


In [13]:
# Optional solution, just print this out and then type in 2774 for the answer below
df_imputed['Garage Qual'].value_counts(dropna=False)

Garage Qual
TA    2774
Fa     124
Gd      24
Po       5
Ex       3
Name: count, dtype: int64

In [14]:
# Set this variable to the number of occurrences of the category 'TA' in the feature 'Garage Qual'
# It should have increased from before the imputation, because Nan values were changed to 'TA'

a1b = df_imputed['Garage Qual'].value_counts(dropna=False).loc['TA']      

print(f"a1b = {a1b}")                                   # Don't delete or change this line, it is needed by the auto-grader

a1b = 2774


### Part C:  Imputing Numeric Features using the Median

Now you must "simply impute" values for the numeric features using the `median` strategy. 

In [15]:

num_imputer = SimpleImputer(strategy="median")

df_imputed[numeric_features] = num_imputer.fit_transform(df_imputed[numeric_features])

# Verify: There should be no null values

show_null_counts_features(df_imputed)

The dataset contains 2930 samples.

There are no null values in the dataset!


In [16]:
# Nothing to do here:   Answer should be 0 

a1c = df_imputed.isnull().any(axis=1).sum()      # count number of rows with any missing values

print(f"a1c = {a1c}")                            # Don't delete or change this line, it is needed by the auto-grader

a1c = 0


### Part D:   Ordinal Encoding the categorical features with OrdinalEncoder

The simplest option in dealing with categorical values is to represent them by integers 0, 1, 2, etc.

**Before proceeding, read the doc page on `sklearn`'s `OrdinalEncoder`.**

Follow the comments to perform this feature transformation.

#### Note: Do not simply apply the OrdinalEncoder to the entire dataframe, as it will also encode the numeric features!
#### You can solve this using the same approach as the transformation in the previous problem. 

In [17]:
# Put df_imputed in the form X, y

X = df_imputed.drop(columns=['SalePrice'])               # Important to note that this makes a copy of df_imputed
y = df_imputed['SalePrice'] 

# Initialize OrdinalEncoder
ordinal_encoder = OrdinalEncoder()

X[categorical_features] = ordinal_encoder.fit_transform(X[categorical_features])

# In this case, the conversion back to a dataframe is unnecessary. 
# # Convert back to DataFrame to retain column names 
# X = pd.DataFrame(X_temp, columns=X.columns, index=X.index)

# # Verify

X.head()

MS SubClass  MS Zoning  Lot Frontage  Lot Area  Street  Lot Shape  \
0         20.0        5.0         141.0   31770.0     1.0        0.0   
1         20.0        4.0          80.0   11622.0     1.0        3.0   
2         20.0        5.0          81.0   14267.0     1.0        0.0   
3         20.0        5.0          93.0   11160.0     1.0        3.0   
4         60.0        5.0          74.0   13830.0     1.0        0.0   

   Land Contour  Utilities  Lot Config  Land Slope  ...  Open Porch SF  \
0           3.0        0.0         0.0         0.0  ...           62.0   
1           3.0        0.0         4.0         0.0  ...            0.0   
2           3.0        0.0         0.0         0.0  ...           36.0   
3           3.0        0.0         0.0         0.0  ...            0.0   
4           3.0        0.0         4.0         0.0  ...           34.0   

   Enclosed Porch  3Ssn Porch  Screen Porch  Pool Area  Misc Val  Mo Sold  \
0             0.0         0.0           0.0        0.0       0.0      5.0   
1             0.0         0.0         120.0        0.0       0.0      6.0   
2             0.0         0.0           0.0        0.0   12500.0      6.0   
3             0.0         0.0           0.0        0.0       0.0      4.0   
4             0.0         0.0           0.0        0.0       0.0      3.0   

   Yr Sold  Sale Type  Sale Condition  
0   2010.0        9.0             4.0  
1   2010.0        9.0             4.0  
2   2010.0        9.0             4.0  
3   2010.0        9.0             4.0  
4   2010.0        9.0             4.0  

[5 rows x 73 columns]

In [18]:
# Nothing to do here:   Answer should show categories encoded as floats for 'Lot Shape'

a1d = X['Lot Shape'].unique()                      

print(f"a1d = {a1d}")                            # Don't delete or change this line, it is needed by the auto-grader

a1d = [0. 3. 1. 2.]


## Problem Two:  Train and Test a Regression Model with Cross-Validation

In this problem, we will perform a regression on the Ames Housing Dataset using several different cross-validation
strategies, comparing the cross-validation score for each with the final testing MSE, to see which provides the best
estimate of the final test score, and hence of the model's ability to generalize. 


We shall compare each of the following cross-validation MSEs with the final test MSE score:

- 5-Fold Cross-Validation (default)
- Repeated 5-Fold Cross-Validation (repeated 100 times)
- Leave-One-Out Cross Validation


### Part A: 5-Fold Cross-Validation

For this part
- Create a train-test split with `test_size=0.2` and `random_state=42`
- Create a linear model and perform K-fold cross-validation with K = 5 and using `scoring='neg_mean_squared_error'` (remember to take the mean of the CV scores and negate the result, since scoring uses a negative MSE). 
- Report the
    - CV score (negated mean of MSE measurements over all K folds)
    - Test MSE


Use `random_state = 42` for all experiments. 

In [19]:
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


# Save the datasets to CSV files
X_train.to_csv("X_train.csv", index=False)
X_test.to_csv("X_test.csv", index=False)
y_train.to_csv("y_train.csv", index=False)
y_test.to_csv("y_test.csv", index=False)

print("Files saved: X_train.csv, X_test.csv, y_train.csv, y_test.csv")


# Initialize the Linear Regression model
model = LinearRegression()

# Perform cross-validation on the training set (using 5-fold CV)
cv_mse_a = -cross_val_score(model, X_train, y_train, cv=5, scoring='neg_mean_squared_error').mean()

# Train the model on the full training set
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate performance on the test set
test_mse = mean_squared_error(y_test, y_pred)

# Print results
print(f"Mean of CV Scores:    {cv_mse_a:.4f}")

print(f"Test RMSE: {test_mse:.4f}")

Files saved: X_train.csv, X_test.csv, y_train.csv, y_test.csv
Mean of CV Scores:    991331210.4187
Test RMSE: 1118604729.6597


In [20]:
# Assign to this variable the Test MSE 

a2a = test_mse                

print(f"a2a = {a2a:.4f}")                            # Don't delete or change this line, it is needed by the auto-grader

a2a = 1118604729.6597


## Part B: Perform Repeated 5-Fold Cross Validation

Read the doc page on `sklearn`'s `RepeatedKFold` for cross validation; repeat the CV calculation with K = 5 and `n_repeats=100`
and report the CV score (negated mean of MSE measurements over all 100*K folds)

In [21]:
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize the Linear Regression model
model = LinearRegression()

# Define Repeated K-Fold Cross-Validation
cv = RepeatedKFold(n_splits=5, n_repeats=100, random_state=42)

# Perform cross-validation on the training set (using 5-fold CV)
cv_mse_b = -cross_val_score(model, X_train, y_train, cv=cv, scoring='neg_mean_squared_error').mean()

# Print results
print(f"Mean of CV Scores:    {cv_mse_b:.4f}")


Mean of CV Scores:    1077441553.3662


In [22]:
# Assign to this variable the mean CV score for the repeated K-fold experiment 
# Note: if your CV score is negative, go back and read the instructions for Part A again

a2b = cv_mse_b               

print(f"a2b = {a2b:.4f}")                            # Don't delete or change this line, it is needed by the auto-grader

a2b = 1077441553.3662


## Part C: Perform Leave One Out Cross Validation

This is simply a matter of setting `cv=LeaveOneOut()`. Run the same experiment and report the CV score. 

In [23]:
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize the Linear Regression model
model = LinearRegression()

# Define LOOCV
loo = LeaveOneOut()

# Perform cross-validation on the training set (using 5-fold CV)
cv_mse_c = -cross_val_score(model, X_train, y_train, cv=loo, scoring='neg_mean_squared_error').mean()

# Print results
print(f"Mean of CV Scores:    {cv_mse_c:.4f}")



Mean of CV Scores:    1053535727.6073


In [24]:
# Assign to this variable the mean CV score for the leave-one-out experiment 

a2c = cv_mse_c               

print(f"a2c = {a2c:.4f}")                            # Don't delete or change this line, it is needed by the auto-grader

a2c = 1053535727.6073


### Part D

Now, in order to help interpret the results, print out a table of the **square roots** of each of the CV scores and the final test score; we can then see the result in the same units (dollars) as the target, instead of the units of the MSE (dollars squared). 

Hint: Here is an example of how to print out values as currency:

    cost = 23512.23
    print(f"cost: ${cost:,.2f}")

In [25]:
print(f"A: RMSE for 5-Fold CV:          ${cv_mse_a**0.5:,.2f}")
print(f"B: RMSE for Repeated 5-Fold CV: ${cv_mse_b**0.5:,.2f}")
print(f"C: RMSE for LOOCV:              ${cv_mse_c**0.5:,.2f}")
print(f"Test RMSE:                      ${test_mse**0.5:,.2f}")  

A: RMSE for 5-Fold CV:          $31,485.41
B: RMSE for Repeated 5-Fold CV: $32,824.40
C: RMSE for LOOCV:              $32,458.22
Test RMSE:                      $33,445.55


In [26]:
# Assign to this variable the letter of the CV strategy whose RMSE came closest to the actual test score

a2d = 'B'                       # Should be 'A' = 5-fold CV; 'B' = Repeated 5-Fold CV; or 'C' = LOO CV                   

print(f"a2d = {a2d}")         # Don't delete or change this line, it is needed by the auto-grader

a2d = B


In [27]:
# How many hours did you spend on this homework?  Assign to the variable an integer value. 

a2e = 5

print(f'a2e = {a2e}')             # Do NOT change this print statement in any way, it is used by the auto-grader!

a2e = 5


## Appendix: Explanation of Features in Ames Housing Dataset

### **Identification**
- `PID` → Parcel Identification Number (unique identifier for each property)
- `Order` → Row number (used for indexing, not a feature)

---

### **Sale Information**
- `SalePrice` → The final selling price of the house in USD (**Target variable**)
- `Mo Sold` → Month the house was sold (1 = January, ..., 12 = December)
- `Yr Sold` → Year the house was sold
- `Sale Type` → Type of sale (e.g., **WD** = Warranty Deed, **New** = Newly Built)
- `Sale Condition` → Condition of the sale (e.g., **Normal**, **Abnormal**, **Partial** for incomplete homes)

---

### **General Property Information**
- `MS SubClass` → Type of dwelling (e.g., **20 = 1-story**, **60 = 2-story**, **120 = Townhouse**)
- `MS Zoning` → Zoning classification (e.g., **RL = Residential Low Density**, **C = Commercial**)
- `Lot Frontage` → Linear feet of street connected to property
- `Lot Area` → Total size of the lot in square feet
- `Neighborhood` → Physical locations within Ames (e.g., **CollgCr = College Creek**)
- `Condition 1` / `Condition 2` → Proximity to roads or railroads (e.g., **Norm = Normal**, **PosN = Near Park**)

---

### **Building & House Design**
- `Bldg Type` → Type of dwelling (e.g., **1Fam = Single Family**, **Twnhs = Townhouse**)
- `House Style` → Style of the house (e.g., **1Story = One Story**, **2Story = Two Story**, **SplitFoyer**)
- `Overall Qual` → Overall quality of materials (scale: **1 = Very Poor** to **10 = Excellent**)
- `Overall Cond` → Overall condition of the house (scale: **1 = Very Poor** to **10 = Excellent**)

---

### **Year Built & Remodel**
- `Year Built` → Original construction year
- `Year Remod/Add` → Year of last remodel or addition

---

### **Exterior Features**
- `Exterior 1st` / `Exterior 2nd` → Exterior covering material (e.g., **VinylSd = Vinyl Siding**, **HdBoard = Hardboard**)
- `Mas Vnr Type` → Masonry veneer type (e.g., **BrkFace = Brick Face**, **None = No Veneer**)
- `Mas Vnr Area` → Area of masonry veneer in square feet

---

### **Basement Features**
- `Bsmt Qual` → Basement height (e.g., **Ex = Excellent**, **TA = Typical**, **Po = Poor**)
- `Bsmt Cond` → General condition of the basement
- `Bsmt Exposure` → Walkout or garden level basement?
- `BsmtFin Type 1` / `BsmtFin SF 1` → Primary finished area in basement (e.g., **GLQ = Good Living Quarters**)
- `BsmtFin Type 2` / `BsmtFin SF 2` → Secondary finished area
- `Bsmt Unf SF` → Unfinished square feet in basement
- `Total Bsmt SF` → Total square footage of basement

---

### **Utilities & HVAC**
- `Heating` → Type of heating system (e.g., **GasA = Gas Forced Air**, **OthW = Hot Water Heating**)
- `Heating QC` → Quality of heating system (e.g., **Ex = Excellent**, **Fa = Fair**)
- `Central Air` → **Y = Yes**, **N = No**
- `Electrical` → Electrical system (e.g., **SBrkr = Standard Breaker**, **FuseA = Fuse Box**)

---

### **Above Ground Living Area**
- `1st Flr SF` → First-floor square footage
- `2nd Flr SF` → Second-floor square footage
- `Gr Liv Area` → Total **above-ground** living area in square feet
- `Low Qual Fin SF` → Low-quality finished square feet (e.g., unfinished rooms)

---

### **Bathrooms & Bedrooms**
- `Full Bath` → Full bathrooms above ground
- `Half Bath` → Half bathrooms above ground
- `Bsmt Full Bath` → Full bathrooms in basement
- `Bsmt Half Bath` → Half bathrooms in basement
- `Bedroom AbvGr` → Number of bedrooms above ground
- `Kitchen AbvGr` → Number of kitchens above ground
- `Kitchen Qual` → Kitchen quality (**Ex = Excellent**, **Fa = Fair**)

---

### **Garage Features**
- `Garage Type` → Type of garage (e.g., **Attchd = Attached**, **Detchd = Detached**)
- `Garage Yr Blt` → Year garage was built
- `Garage Finish` → Interior finish of garage
- `Garage Cars` → Size of garage in car capacity
- `Garage Area` → Garage size in square feet

---

### **Additional Features**
- `Fireplaces` → Number of fireplaces
- `Fireplace Qu` → Fireplace quality
- `Paved Drive` → Paved driveway? (**Y = Yes, P = Partial, N = No**)
- `Wood Deck SF` → Square footage of wood deck
- `Open Porch SF` → Square footage of open porch
- `Enclosed Porch` → Square footage of enclosed porch
- `Screen Porch` → Square footage of screened porch
- `Pool Area` → Pool area in square feet
- `Misc Val` → Miscellaneous features (e.g., shed value)

